File: forge/agent/state.py
"""Agent states, lifecycle events, and message data structures."""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional
import time


class AgentState(str, Enum):
    IDLE = "idle"
    THINKING = "thinking"
    PLANNING = "planning"
    READING = "reading"
    EXECUTING = "executing"
    TESTING = "testing"
    WAITING_APPROVAL = "waiting_approval"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class ToolExecutionResult:
    tool_name: str
    arguments: Dict[str, Any]
    output: str
    success: bool = True
    duration_ms: int = 0


@dataclass
class AgentEvent:
    type: str  # "state_changed", "text_chunk", "text_output", "tool_start", "tool_end", "approval_needed", "done", "error"
    data: Any = None
    timestamp: float = field(default_factory=time.time)

    @classmethod
    def state_changed(cls, state: AgentState) -> "AgentEvent":
        return cls("state_changed", state)

    @classmethod
    def text_chunk(cls, text: str) -> "AgentEvent":
        return cls("text_chunk", text)

    @classmethod
    def text_output(cls, text: str) -> "AgentEvent":
        return cls("text_output", text)

    @classmethod
    def tool_start(cls, name: str, arguments: dict) -> "AgentEvent":
        return cls("tool_start", {"name": name, "arguments": arguments})

    @classmethod
    def tool_end(cls, result: ToolExecutionResult) -> "AgentEvent":
        return cls("tool_end", result)

    @classmethod
    def approval_needed(cls, tool_name: str, arguments: dict, warning: Optional[str] = None) -> "AgentEvent":
        return cls("approval_needed", {"tool_name": tool_name, "arguments": arguments, "warning": warning})

    @classmethod
    def done(cls, final_text: str = "") -> "AgentEvent":
        return cls("done", final_text)

    @classmethod
    def error(cls, message: str) -> "AgentEvent":
        return cls("error", message)

File: archive/rust_legacy/src/agent/state.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum AgentState {
    Idle,
    Thinking,
    Planning,
    Reading,
    Executing,
    Testing,
    WaitingApproval,
    Completed,
    Failed(String),
}

impl AgentState {
    pub fn symbol(&self) -> &str {
        match self {
            AgentState::Idle => "●",
            AgentState::Thinking => "◉",
            AgentState::Planning => "◉",
            AgentState::Reading => "◉",
            AgentState::Executing => "▶",
            AgentState::Testing => "◉",
            AgentState::WaitingApproval => "◆",
            AgentState::Completed => "✓",
            AgentState::Failed(_) => "✗",
        }
    }

    pub fn label(&self) -> String {
        match self {
            AgentState::Idle => "Ready".to_string(),
            AgentState::Thinking => "Thinking...".to_string(),
            AgentState::Planning => "Planning...".to_string(),
            AgentState::Reading => "Reading repository...".to_string(),
            AgentState::Executing => "Executing...".to_string(),
            AgentState::Testing => "Running tests...".to_string(),
            AgentState::WaitingApproval => "Waiting for approval".to_string(),
            AgentState::Completed => "Completed".to_string(),
            AgentState::Failed(msg) => format!("Failed: {}", msg),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub input: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResult {
    pub tool_name: String,
    pub success: bool,
    pub output: String,
    pub duration_ms: u64,
}

#[derive(Debug, Clone)]
pub enum AgentEvent {
    StateChanged(AgentState),
    TextOutput(String),
    ToolStarted(ToolCall),
    ToolCompleted(ToolResult),
    ApprovalRequired { command: String, description: String },
    Error(String),
    Done,
}

/// A message in the conversation history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationMessage {
    pub role: MessageRole,
    pub content: String,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MessageRole {
    User,
    Agent,
    Tool,
    System,
}

impl ConversationMessage {
    pub fn user(content: impl Into<String>) -> Self {
        Self {
            role: MessageRole::User,
            content: content.into(),
            timestamp: chrono::Utc::now(),
        }
    }

    pub fn agent(content: impl Into<String>) -> Self {
        Self {
            role: MessageRole::Agent,
            content: content.into(),
            timestamp: chrono::Utc::now(),
        }
    }

    pub fn tool(content: impl Into<String>) -> Self {
        Self {
            role: MessageRole::Tool,
            content: content.into(),
            timestamp: chrono::Utc::now(),
        }
    }
}
